azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
+                                    status=registration_status, status_code=success_status_code
+                                )
+                            )
+                            initial_register_op.complete(error=error)
+
+                else:
+                    initial_register_op.complete(error=error)
+
+            registration_payload = DeviceRegistrationPayload(
+                registration_id=initial_register_op.registration_id,
+                custom_payload=initial_register_op.request_payload,
+            )
+            self.send_op_down(
+                pipeline_ops_base.RequestAndResponseOperation(
+                    request_type=constant.REGISTER,
+                    method="PUT",
+                    resource_location="/",
+                    request_body=registration_payload.get_json_string(),
+                    callback=on_registration_response,
+                )
+            )
+
+        else:
+            super(RegistrationStage, self)._run_op(operation)
+
+
+class ProvisioningTimeoutStage(PipelineStage):
+    """
+    The purpose of the timeout stage is to add timeout errors to select operations
+
+    The timeout_intervals attribute contains a list of operations to track along with
+    their timeout values.
+
+    For each operation that needs a timeout check, this stage will add a timer to
+    the operation.  If the timer elapses, this stage will fail the operation with
+    a PipelineTimeoutError.
+    """
+
+    def __init__(self):
+        super(ProvisioningTimeoutStage, self).__init__()
+        self.timeout_intervals = {
+            pipeline_ops_base.RequestOperation: constant.DEFAULT_TIMEOUT_INTERVAL
+        }
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, op):
+        if type(op) in self.timeout_intervals:
+            # Create a timer to watch for operation timeout on this op and attach it
+            # to the op.
+            self_weakref = weakref.ref(self)
+
+            @pipeline_thread.invoke_on_pipeline_thread_nowait
+            def on_timeout():
+                this = self_weakref()
+                logger.info("{}({}): returning timeout error".format(this.name, op.name))
+                # For DPS it is an Service Timeout Error as opposed to pipeline timing out
+                op.complete(
+                    error=(
+                        exceptions.ServiceError(
+                            "Operation timed out before provisioning service could respond for {type_of} operation".format(
+                                type_of=op.request_type
+                            )
+                        )
+                    )
+                )
+
+            logger.debug("{}({}): Creating timer".format(self.name, op.name))
+            op.timeout_timer = Timer(constant.DEFAULT_TIMEOUT_INTERVAL, on_timeout)
```

Comment:
>timeout_timer [](start = 15, length = 13)

Also this is the same timer variable name used in the `OpTimeoutStage`. While I don't think any single operation can trigger both stages, I still think it's probably best if we can give this a separate name.

We also really ought to instantiate all these attributes initially (as None) on all operations via the parent Operation object, because dynamically applying them is asking for trouble. However, that's a change that spans both pipelines and all operations, so best to save it for a different PR.

Reply:
resolved for first one.

---
In reply to: [364530593](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r364530593) [](ancestors = 364530593)